File: domain/src/main/scala/net/kemitix/thorp/domain/RemoteKey.scala
package net.kemitix.thorp.domain

import java.io.File
import java.nio.file.{Path, Paths}
import Implicits._

final case class RemoteKey(key: String)

object RemoteKey {
  val key: SimpleLens[RemoteKey, String] =
    SimpleLens[RemoteKey, String](_.key, b => a => b.copy(key = a))
  def asFile(source: Path, prefix: RemoteKey)(
      remoteKey: RemoteKey): Option[File] =
    if (remoteKey.key.length === 0) None
    else Some(source.resolve(RemoteKey.relativeTo(prefix)(remoteKey)).toFile)
  def relativeTo(prefix: RemoteKey)(remoteKey: RemoteKey): Path = {
    prefix match {
      case RemoteKey("") => Paths.get(remoteKey.key)
      case _             => Paths.get(prefix.key).relativize(Paths.get(remoteKey.key))
    }
  }
  def resolve(path: String)(remoteKey: RemoteKey): RemoteKey =
    RemoteKey(List(remoteKey.key, path).filterNot(_.isEmpty).mkString("/"))
  def fromSourcePath(source: Path, path: Path): RemoteKey = {
    RemoteKey(source.relativize(path).toString)
  }
}

File: domain/src/main/java/net/kemitix/thorp/domain/RemoteKey.java
package net.kemitix.thorp.domain;

import net.kemitix.mon.TypeAlias;

import java.io.File;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class RemoteKey extends TypeAlias<String> {
    private RemoteKey(String value) {
        super(value);
    }
    public static RemoteKey create(String key) {
        return new RemoteKey(key);
    }
    public String key() {
        return getValue();
    }
    public Optional<File> asFile(Path source, RemoteKey prefix) {
        if (key().length() == 0 || !key().startsWith(prefix.key())) {
            return Optional.empty();
        }
        return Optional.of(
                source.resolve(relativeTo(prefix))
                        .toFile());
    }
    public Path relativeTo(RemoteKey prefix) {
        if (prefix.key().equals("")) {
            return Paths.get(key());
        }
        return Paths.get(prefix.key()).relativize(Paths.get(key()));
    }
    public RemoteKey resolve(String path) {
        return RemoteKey.create(
                Stream.of(key(), path)
                        .filter(s -> !s.isEmpty())
                        .collect(Collectors.joining("/")));
    }
    public static RemoteKey fromSourcePath(Path source, Path path) {
        return RemoteKey.create(
                source.relativize(path).toString());
    }
    public static RemoteKey from(Path source, RemoteKey prefix, File file) {
        return prefix.resolve(source.relativize(file.toPath()).toString());
    }
}
